internal/linker/index.js

```diff
@@ -393,28 +365,73 @@ function visitDirectoryPreserveLinks(dirPath, visit) {
-        log_verbose('resolved node_modules root', root, 'to', rootDir);
-        log_verbose('cwd', process.cwd());
-        if (!(yield exists(rootDir))) {
-            log_verbose('no third-party packages; mkdir node_modules at', root);
-            yield mkdirp(rootDir);
-        }
-        yield symlink(rootDir, 'node_modules');
-        process.chdir(rootDir);
+        log_verbose('startCwd:', startCwd);
+        const execroot = findExecroot(startCwd);
+        log_verbose('execroot:', execroot ? execroot : 'not found');
+        const isExecroot = startCwd == execroot;
+        log_verbose('isExecroot:', isExecroot.toString());
+        if (!isExecroot && execroot) {
+            process.chdir(execroot);
+            log_verbose('changed directory to execroot', execroot);
+        }
+        for (const packagePath of Object.keys(roots)) {
+            const workspace = roots[packagePath];
+            const workspacePath = yield resolveExternalWorkspacePath(workspace, startCwd, isExecroot, execroot, runfiles);
+            log_verbose(`resolved ${workspace} workspace path to ${workspacePath}`);
+            const workspaceNodeModules = `${workspacePath}/node_modules`;
+            if (packagePath) {
+                if (yield exists(workspaceNodeModules)) {
+                    let resolvedPackagePath;
+                    if (!isExecroot) {
+                        const runfilesPackagePath = `${startCwd}/${packagePath}`;
+                        if (yield exists(runfilesPackagePath)) {
+                            yield symlink(workspaceNodeModules, `${runfilesPackagePath}/node_modules`);
+                            resolvedPackagePath = runfilesPackagePath;
+                        }
+                    }
+                    if (yield exists(packagePath)) {
+                        yield symlink(workspaceNodeModules, `${packagePath}/node_modules`);
+                        resolvedPackagePath = packagePath;
+                    }
+                    const packagePathBin = `${bin}/${packagePath}`;
+                    if (resolvedPackagePath && (yield exists(packagePathBin))) {
+                        yield symlink(`${resolvedPackagePath}/node_modules`, `${packagePathBin}/node_modules`);
+                    }
+                }
+            }
+            else {
+                if (yield exists(workspaceNodeModules)) {
+                    yield symlink(workspaceNodeModules, `node_modules`);
+                }
+                else {
+                    log_verbose('no root npm workspace node_modules folder to link to; creating node_modules directory in', process.cwd());
+                    yield mkdirp('node_modules');
+                }
+            }
+        }
+        const rootNpmWorkspace = roots ? roots[''] : undefined;
+        if (!rootNpmWorkspace) {
+            log_verbose('no root npm workspace; creating node_modules directory in ', process.cwd());
+            yield mkdirp('node_modules');
```

Comment:
Should this also set the `rootNpmWorkspace` since I assume you are now creating one?

Reply:
Not used anywhere else.
Changed to
```
if (!roots || !roots['']) {
```
to be more clear on that